openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+
+        # Ensure email is set in the session
+        self.assertEqual(self.request.session.get('email'), self.user.email)
+        # Ensure session cookie exist
+        self.assertEqual(len(self.client.response.cookies), 1)
+
+        # simulating email changed
+        self.user.email = 'new_email@test.com'
+        self.user.save()
+
+        # Call process_request
+        EmailChangeMiddleware(get_response=lambda request: None).process_request(self.request)
+
+        # Verify that session_email_match and is_enforce_session_email_match_enabled
+        # custom attributes are set
+        mock_set_custom_attribute.assert_has_calls([call('session_email_match', False)])
+        mock_set_custom_attribute.assert_has_calls([call('is_enforce_session_email_match_enabled', True)])
+
+        # Assert that the session is flushed and cookies marked for deletion
+        mock_mark_cookie_for_deletion.assert_called()
+        assert self.request.session.get(SESSION_KEY) is None
+        assert self.request.user == AnonymousUser()
+
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_logged_in_cookies")
+    def test_process_response_user_not_authenticated(self, mock_set_logged_in_cookies):
+        """
+        Calls EmailChangeMiddleware.process_response when user is not authenticated.
+        Verify that the logged-in cookies are not updated
+        """
+        # return value of mock
+        mock_set_logged_in_cookies.return_value = self.client.response
+
+        # Unauthenticated User
+        self.request.user = AnonymousUser()
+
+        # Call process_response without authenticating a user
+        response = EmailChangeMiddleware(get_response=lambda request: None).process_response(
+            self.request, self.client.response
+        )
+
+        assert response.status_code == 200
+        # Assert that cookies are not updated
+        # Assert that mock_set_logged_in_cookies not called
+        mock_set_logged_in_cookies.assert_not_called()
+
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_logged_in_cookies")
+    def test_process_response_email_change_requested(self, mock_set_logged_in_cookies):
+        """
+        Calls EmailChangeMiddleware.process_response when user is authenticated and email
+        change was requested.
+        Verify that the logged-in cookies are updated
+        """
+        # return value of mock
+        mock_set_logged_in_cookies.return_value = self.client.response
+
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.TEST_PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (simulating email change requested)
+        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```

Comment:
1. Could this use the following instead (copied from another test, with updated comment)?
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
```
2. I think any test call to `register_email_change` that makes sense to replace with actual user save code (like above), will make a more resilient test.

Reply:
> Could this use the following instead (copied from another test, with updated comment)?

The `process_response` function of middleware relies on the `email_change_requested` variable set in the request cache and it is set in the `register_email_change` function of middleware.

Therefore, `register_email_change` should be called here to set the `email_change_requested` variable. 

But we can make the following change by having both the email change simulation and email change registration. It can make more sense?
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
        # Registering email change (simulating email change requested)
        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```

Please let me know your thoughts about this.

-------------------------------------------------------------------------------------------------

> I think any test call to register_email_change that makes sense to replace with actual user save code (like above), will make a more resilient test.

There were other uses of `register_email_change` in `test_process_request_emails_match` and `test_process_request_emails_mismatch` tests.  There, the `register_email_change` function was being used for setting user email in session for later comparison by middleware's `proces_request` function. For now I have kept them but we can also remove them and can set the email in session ourselves. as below

```
self.request.session['email'] = self.user.email  # setting email in session for comparison
```

Please let me know your thoughts on this.

Reply:
From above:    
```
        # simulating new email change
        self.user.email = 'new_email@test.com'
        self.user.save()
        # Registering email change (simulating email change requested)
        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
```
> Please let me know your thoughts about this.

The email save doesn't trigger the signal receiver that already calls `register_email_change`?

> There, the `register_email_change` function was being used for setting user email in session for later comparison by middleware's `process_request `function.

Is this where the comment `Simulates a past email change (e.g. from an earlier request)` (or something like that) would apply? If so, I get why it is needed and I'd just like an appropriate comment to clarify why it was needed.

Reply:
> The email save doesn't trigger the signal receiver that already calls register_email_change?

Yes this happens in real situation but the signal receiver is not called with `self.user.save()` so we have to call `register_email_change` explicitly in tests.
But I am trying to write integration tests and hopefully that flow will execute in those tests.

Reply:
Marking resolved.